File: holoeasy-core/src/main/java/org/holoeasy/line/LineImpl.java
package org.holoeasy.line;

import org.bukkit.Location;
import org.bukkit.entity.EntityType;
import org.bukkit.entity.Player;
import org.bukkit.scheduler.BukkitTask;
import org.holoeasy.animation.Animations;
import org.holoeasy.hologram.Hologram;
import org.jetbrains.annotations.ApiStatus;
import org.jetbrains.annotations.NotNull;

import java.util.Random;
import java.util.concurrent.atomic.AtomicInteger;

public abstract class LineImpl<T> implements Line<T> {

    protected final Hologram hologram;
    protected final EntityType entityType;
    private Location location;

    public abstract @NotNull Type getType();

    public final int entityID;

    @Override
    public int getEntityID() {
        return entityID;
    }

    @Override
    public abstract @NotNull T getValue();

    public abstract void setValue(T value);

    private BukkitTask animationTask;

    public LineImpl(Hologram hologram, EntityType entityType) {
        this.hologram = hologram;
        this.entityType = entityType;
        this.entityID = IDs_COUNTER.getAndIncrement();
    }

    public Hologram getHologram() {
        return hologram;
    }

    @Override
    public Location currentLocation() {
        return location;
    }

    @Override
    public abstract void update(@NotNull T newValue);

    @Override
    public void setAnimation(Animations animation) {
        cancelAnimation();
        animationTask = animation.getTask().apply(this);
    }

    @Override
    public void cancelAnimation() {
        if (animationTask != null) {
            animationTask.cancel();
            animationTask = null;
        }
    }

    public abstract void show(Player player);

    public abstract void hide(Player player);

    public abstract void update(Player player);

    public void destroy(Player player) {
        hologram.getLib().getPacketImpl().deletePacket(player, entityID);
    }

    public boolean spawn(Player player) {
        Location loc = location;
        if (loc == null) return false;
        hologram.getLib().getPacketImpl().spawn(hologram.getLib(), player, entityID, entityType, loc);
        return true;
    }

    public boolean teleport(Player player) {
        Location loc = location;
        if (loc == null) return false;
        hologram.getLib().getPacketImpl().teleport(player, entityID, loc);
        return true;
    }

    public void setCurrentLocation(Location value) {
        this.location = value;
    }

    public void observerUpdate() {
        for (Player player : hologram.getPvt().getSeeingPlayers()) {
            update(player);
        }
    }

    public static final AtomicInteger IDs_COUNTER = new AtomicInteger(500 + new Random().nextInt());

    public enum Type {
        EXTERNAL,
        TEXT_LINE,
        CLICKABLE_TEXT_LINE,
        ITEM_LINE,
        BLOCK_LINE,
        @ApiStatus.Experimental
        DISPLAY_BLOCK_LINE,
        @ApiStatus.Experimental
        DISPLAY_TEXT_LINE
    }
}

File: holoeasy-core/src/main/kotlin/org/holoeasy/line/LineImpl.kt
package org.holoeasy.line

import org.bukkit.Location
import org.bukkit.entity.EntityType
import org.bukkit.entity.Player
import org.bukkit.scheduler.BukkitTask
import org.holoeasy.animation.Animations
import org.holoeasy.hologram.Hologram
import org.jetbrains.annotations.ApiStatus
import java.util.*
import java.util.concurrent.atomic.AtomicInteger

sealed class LineImpl<T>(
    val hologram: Hologram,
    val entityType: EntityType
) : Line<T> {

 private var location: Location? = null

    // Exposed Methods

    abstract override val type: Type

    override val entityID: Int = IDs_COUNTER.getAndIncrement()

    // abstract internal
    abstract override var value: T

    override fun currentLocation() = location

    abstract override fun update(newValue: T)


    private var animationTask: BukkitTask? = null

    override fun setAnimation(animation: Animations) {
        this.cancelAnimation()
        animationTask = animation.task(this)
    }

    override fun cancelAnimation() {
        animationTask?.cancel()
        animationTask = null
    }

    // Internal

    abstract fun show(player: Player)


    abstract fun hide(player: Player)


    abstract fun update(player: Player)


    fun destroy(player: Player) {
        hologram.lib.packetImpl.deletePacket(player, entityID)
    }


    fun spawn(player: Player): Boolean {
        hologram.lib.packetImpl.spawn(hologram.lib, player, entityID, entityType, location ?: return false)
        return true
    }


    fun teleport(player: Player): Boolean {
        hologram.lib.packetImpl.teleport(player, entityID, location ?: return false)
        return true
    }


    open fun setCurrentLocation(value: Location) {
        this.location = value
    }


    fun observerUpdate() {
        for (player in hologram.pvt.seeingPlayers) {
            update(player)
        }
    }

    companion object {
        val IDs_COUNTER = AtomicInteger(500 + Random().nextInt())
    }

    enum class Type {
        EXTERNAL,

        TEXT_LINE,
        CLICKABLE_TEXT_LINE,
        ITEM_LINE,
        BLOCK_LINE,

        @ApiStatus.Experimental
        DISPLAY_BLOCK_LINE,
        @ApiStatus.Experimental
        DISPLAY_TEXT_LINE
    }


}